File: src/tree_algorithms/tree_distance_i.py
import sys


def input():
    return sys.stdin.readline()


def print(x):
    return sys.stdout.write(str(x) + " ")


sys.setrecursionlimit(10**9)
n = int(input())

adjList = [[] for _ in range(n)]

for _ in range(n - 1):
    i, j = map(int, input().split())
    i, j = i - 1, j - 1
    adjList[i].append(j)
    adjList[j].append(i)

max_depth1 = [None for _ in range(n)]
max_depth2 = [None for _ in range(n)]

ans = 0


def dfs1(node, parent):
    d1, d2 = 0, 0
    for adj in adjList[node]:
        if adj != parent:
            v = dfs1(adj, node)[0] + 1
            if v >= d1:
                d1, d2 = v, d1
            elif v > d2:
                d2 = v
    max_depth1[node] = d1, d2
    # print(node + 1, parent + 1, d)
    return max_depth1[node]


def dfs2(node, parent):
    parentMaxL = -1
    if parent is not None:
        if max_depth1[node][0] == max_depth1[parent][0] - 1:
            parentMaxL = max_depth1[parent][1] + 1
        else:
            parentMaxL = max_depth1[parent][0] + 1
    if parentMaxL >= max_depth1[node][0]:
        max_depth1[node] = parentMaxL, max_depth1[node][0]
    elif parentMaxL >= max_depth1[node][1]:
        max_depth1[node] = max_depth1[node][0], parentMaxL

    max_depth2[node] = max_depth1[node][0]
    # print(node, max_depth1)
    for adj in adjList[node]:
        if adj != parent:
            dfs2(adj, node)


dfs1(0, None)
dfs2(0, None)

# print(max_depth1)
print(" ".join(map(str, max_depth2)))

File: src/tree_algorithms/tree_distance_i.rs
use std::io;

fn compute_max_depth1(
    node: usize,
    parent: Option<usize>,
    adj_list: &Vec<Vec<usize>>,
    max_depth: &mut Vec<(i32, i32)>,
    ans: &mut i32,
) -> (i32, i32) {
    let mut m1 = 0;
    let mut m2 = 0;

    for &adj in &adj_list[node] {
        if Some(adj) != parent {
            let (v1, _v2) = compute_max_depth1(adj, Some(node), adj_list, max_depth, ans);
            let v = v1 + 1;
            if v >= m1 {
                m2 = m1;
                m1 = v;
            } else if v > m2 {
                m2 = v;
            }
        }
    }

    *ans = (*ans).max(m1 + m2);
    max_depth[node] = (m1, m2);
    (m1, m2)
}

fn compute_max_depth2(
    node: usize,
    parent: Option<usize>,
    adj_list: &Vec<Vec<usize>>,
    max_depth: &mut Vec<(i32, i32)>,
    ans: &mut i32,
) {
    if parent.is_some() {
        let parent_max_l;

        let (m1p,  m2p) = max_depth[parent.unwrap()];
        let (mut m1, mut m2) = max_depth[node];

        if m1p == m1 + 1 {
            parent_max_l = m2p + 1;
        } else {
            parent_max_l = m1p + 1;
        }

        if parent_max_l >= m1 {
            m2 = m1;
            m1 = parent_max_l;
        } else if parent_max_l > m2 {
            m2 = parent_max_l;
        }
        max_depth[node] = (m1, m2)
    }

    for &adj in &adj_list[node] {
        if Some(adj) != parent {
            compute_max_depth2(adj, Some(node), adj_list, max_depth, ans);
        }
    }
}

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read input");
    let n: usize = input.trim().parse().expect("Invalid input");

    let mut adj_list = vec![Vec::new(); n];

    for _ in 0..n - 1 {
        let mut input = String::new();
        io::stdin()
            .read_line(&mut input)
            .expect("Failed to read input");
        let nums: Vec<usize> = input
            .split_whitespace()
            .map(|x| x.parse().expect("Invalid input"))
            .collect();

        let i = nums[0] - 1;
        let j = nums[1] - 1;

        adj_list[i].push(j);
        adj_list[j].push(i);
    }

    let mut max_depth = vec![(-1, -1); n];
    let mut ans = 0;

    compute_max_depth1(0, None, &adj_list, &mut max_depth, &mut ans);
    // println!("{:?}", max_depth);
    compute_max_depth2(0, None, &adj_list, &mut max_depth, &mut ans);
    // println!("{:?}", max_depth);
    println!(
        "{}",
        max_depth
            .into_iter()
            .map(|x| x.0.to_string())
            .collect::<Vec<String>>()
            .join(" ")
    );
}
